classes/Search.php

```diff
@@ -563,11 +565,43 @@ protected static function fillProductArray(&$product_array, $weight_array, $key,
 	public static function indexation($full = false, $id_product = false)
 	{
 		$db = Db::getInstance();
+		$all_product_shop = false;
 
 		if ($id_product)
 			$full = false;
 
-		if ($full)
+		if (Context::getContext()->shop->getContext() == Shop::CONTEXT_SHOP)
+		{
+			$id_shop = Context::getContext()->shop->getContextShopID();
+
+			$db->execute('DELETE si FROM `'._DB_PREFIX_.'search_index` si
+				INNER JOIN `'._DB_PREFIX_.'product` p ON (p.id_product = si.id_product)
+				'.Shop::addSqlAssociation('product', 'p').'
+				WHERE product_shop.`visibility` IN ("both", "search")
+				AND product_shop.`active` = 1
+				AND p.`id_product` IN (
+					SELECT * FROM (
+					SELECT p.`id_product` FROM `'._DB_PREFIX_.'product` p
+					'.Shop::addSqlAssociation('product', 'p').
+					' WHERE product_shop.`id_shop` = '.$id_shop.
+					') AS subquery
+				)');
```

Comment:
Note the first query is not really efficient, it would be great to add a field 'is_multishop' in the 'product' table to store the fact that a product is multishop or not and then just add an index on it.

Then you'd just have to exec :

```php
$db->execute('DELETE si FROM `'._DB_PREFIX_.'search_index` si
                INNER JOIN `'._DB_PREFIX_.'product` p ON (p.id_product = si.id_product)
                '.Shop::addSqlAssociation('product', 'p').'
                WHERE product_shop.`visibility` IN ("both", "search")
                AND product_shop.`active` = 1 AND p.is_multishop=1');
```